---
title: "RCyjs: programmatic access to the web browser-based network viewer, cytoscape.js"
author: "Paul Shannon"
package: RCyjs
date: "`r Sys.Date()`"
output:
   BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{"RCyjs: programmatic access to the web  browser-based network viewer, cytoscape.js"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style>
.main-container { width: 1200px; max-width:2800px;}
</style>


```{r setup, include = FALSE}
options(width=120)
knitr::opts_chunk$set(
   collapse = TRUE,
   eval=interactive(),
   echo=TRUE,
   comment = "#>"
)
```


# Introduction

RCyjs is a BrowserViz subclass providing interactive access to Cytoscape.js running in your
browser from an R session.  Cytoscape.js \url{http://js.cytoscape.org} provides full-featured network visualization in an HTML5
browser window.  This web-based implementation is related to, and can exchange data with the
desktop version of Cytoscape: \url{http://cytoscape.org}.    See the BrowserViz vignette for
a description of the websocket and JSON techniques used here to connect your R session to your browser.

# Simple Example

RCyjs provides a utility function which creates a small 3-node, 3-edge graphNEL (see \Biocpkg{graph})
with some attributes assigned to the nodes and edges:

```{r, eval=TRUE, echo=FALSE}
knitr::include_graphics("igv-vignette.png")
```

```{r simpleExample, results='hide'}
library(RCyjs)
g <- simpleDemoGraph()
noaNames(g)
edaNames(g)
noa(g, "type")
noa(g, "lfc")
eda(g, "edgeType")
eda(g, "score")
```
Send this graph to your browser.  Request a simple layout and apply
some simple rendering rules.

In network visualization we commonly use edge and node attributes to control visual appearance.
For instance, an edgeType of ``phosphorylates'' could be rendered in green, and ``regulates'' in black.
Node size can reflect expression levels or population size.  Node borders can indicate abnormal modifications.
This process, and the rules that govern it, go by the name of ``visual mapping''.

Before discussing and illustrating the use of these rules, let us first see how to set default
visual properties which apply to all parts of the graph.  For instance: make all nodes round, with
white interior, a thin black border, 50 pixels in diameter, connected by black edges (lines) which
have no terminal decoration (that is, no arrows at the ends of the edges).  Note that \emph{redraw}
must be called to \emph{apply} the currently specified rules.  First we define some useful colors:

```{r simpleRender,  results='hide'}
g <- simpleDemoGraph()
rcy <- RCyjs(title=title, graph=g)
setBrowserWindowTitle(rcy, "simple demo")
layout(rcy, "cose")
setGlobalNodeSize(rcy, 30)
setGlobalNodeShape(rcy, "ellipse")
setGlobalNodeColor(rcy,
setGlobalNodeBorderWidth(rcy, 1)
```

# Create a GRanges object specifying our region of interest, 285 kb centered on the MEF2C gene

```{r roi,  results='hide'}
chrom <- "chr5"
shoulder <- 50000
start.loc <- 88013975 - shoulder
end.loc   <- 88199922 + shoulder
mef2c.region <- GRanges(seqnames=chrom, IRanges(start=start.loc, end=end.loc))
showGenomicRegion(igv, list(chrom=chrom, start=start.loc, end=end.loc))
```

# IGAP GWAS: read a tab-delimited file into a data.frame, create and display two tracks

These next few steps expose the nuts and bolts of creating and displaying tracks from tabular data.
None of the power and convenience of the Biocondcutor AnnotationHub, of rtracklayer and
GenomicRanges is used here - that comes later.

Add the IGAP GWAS track twice, first as an annotation track, so that names are avaialble by clicking,
then again as a quantitative track, using a -log10(pValue) score, conveying strength of association
of each SNP with the expression of the target gene, MEF2C.

[IGAP](http://consortiapedia.fastercures.org/consortia/igap/), the International Genomics of Alzheimer's Project, provide these data.  See [Meta-analysis of 74,046 individuals identifies 11 new susceptibility loci for Alzheimerâ€™s disease](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3896259/), Nature Genetics, Dec 2013.


```{r data.frame.track,  results='hide'}
load(system.file(package="igvR", "extdata", "tbl.mef2cGWAS.variants.RData"))
tbl.mef2cGWAS.variants.bed <- tbl.mef2cGWAS.variants[, c("CHR", "oldPos", "oldPos", "SNP", "P")]
tbl.mef2cGWAS.variants.bed$P <- -log10(tbl.mef2cGWAS.variants.bed$P)
colnames(tbl.mef2cGWAS.variants.bed) <- c("chrom", "start", "end", "name", "score")
track.gwas <- DataFrameAnnotationTrack("IGAP.gwas", tbl.mef2cGWAS.variants.bed, trackHeight=20, color="darkBlue")
displayTrack(igv, track.gwas)

tbl.mef2cGWAS.variants.bedGraph <-  tbl.mef2cGWAS.variants.bed[, -4]
track.gwas.numeric <- DataFrameQuantitativeTrack("IGAP.gwas.scored",tbl.mef2cGWAS.variants.bedGraph)
displayTrack(igv, track.gwas.numeric)

```
# Use the AnnotationHub to create a promoter track

A location-only calculation is used to identify gene promoters, rather than more nuanced
experimental data, of which we will seem more later in this vignette.  We use default, conservative
parameters (2kb upstream of TSS, 200bp downstream).

```{r queryAHforPromoters,  results='hide'}
ah <- AnnotationHub()
ah.human <- subset(ah, species == "Homo sapiens")
#----------------------------------------------------------------------------------------------------
# add refseq promoters, available from RefSeq for each transcript which has been identified
#----------------------------------------------------------------------------------------------------
ah.human.refseq <- query(ah.human, "RefSeq", "hg19", "RefSeq Genes")

# download the first set
human.refseq <- ah.human.refseq[[1]]
gr.promoters <- promoters(human.refseq, upstream=2000, downstream=200)
  # get rid of score, itemRgb, thick, blocks columns in the mcols, keeping just the transcript name.
  # these attributes are meaningful for transcript tracks since those include the represenation
  # of UTRs, introns and exons.   but a promoter is a stretch of DNA for which those distinctions
  # do not apply
mcols(gr.promoters) <- mcols(gr.promoters)[,1]
colnames(mcols(gr.promoters)) <- "name"
ov <- findOverlaps(gr.promoters, mef2c.region)
gr.mef2c.promoters <- gr.promoters[queryHits(ov)]
track.promoters <- UCSCBedAnnotationTrack("promoters", gr.mef2c.promoters, color="darkGreen")
displayTrack(igv, track.promoters)
```
# Overlap promoters and variants, create new track

Non-coding SNPs which fall in promoter regions are plausible candidates for further examination.
They may disrupt or create a transcription factor binding site, and thus possibly contribute to the
disease under study.

```{r overlapPromotersAndVariants,  results='hide'}
gr.variants <- GRanges(tbl.mef2cGWAS.variants.bed)
ov <- findOverlaps(gr.variants, gr.promoters)
gr.variantsInPromoters <- gr.variants[queryHits(ov)]
track.variantsInPromoters <-GRangesAnnotationTrack("snpInPromoters", gr.variantsInPromoters,
                                                   color="red", displayMode="EXPANDED")
displayTrack(igv, track.variantsInPromoters)
```
# Add Methylation Tracks

H3K4me3 histone methylation marks are often associated with transcription initiation in promoter
regions.  We will add three tracks from the AnnotationHub for H3K4me3 methylation in
[Gm12878](https://www.genome.gov/26524238/encode-project-common-cell-types/) (lymphoblastoid) cells.

```{r methylationTracks,  results='hide'}
histone.tracks <- query(ah.human, c("H3K4me3", "Gm12878", "Peak", "narrow"))  # 3 tracks
descriptions <- histone.tracks$description
titles <- histone.tracks$title
colors <- rep(terrain.colors(6), 4)
color.index <- 0

for(i in seq_len(length(histone.tracks))){
   name <- names(histone.tracks)[i]
   color.index <- color.index + 1
   gr <- histone.tracks[[name]]
   ov <- findOverlaps(gr, mef2c.region)
   mef2c.histones <- gr[queryHits(ov)]
   track.histones <- GRangesQuantitativeTrack(titles[i], mef2c.histones[, "pValue"],
                                              color=colors[color.index], trackHeight=50)
   displayTrack(igv, track.histones)
   Sys.sleep(5)
   } # for track

```

# VCF from ADNI, the Alzheimer's Disease Neuroimaging Initiative

We include a very small slice of the ADNI whole genome sequencing
[data](http://adni.loni.usc.edu/data-samples/genetic-data/wgs/).  349 samples (filtered from a
larger set of 128 AD, 415 Mild Cognitive Impairment, 267 controls, 8 uncertains) are included, all
in a ~4kb region including methylation marks, and IGAP GWAS SNPs in promoter regions.

In order to illustrate some of genomic data exploration capabilities of igvR we are therefore mixing
apples and oranges.  We have GWAS variants from IGAP (a very large meta-analysis), promoters from
RefSeq (conservatively defined), methylation marks from a lymphoblastoid cell line, and WGS variants
from ADNI. Note also that the ADNI variants in the VCF track are not stratified by disease
phenotype.  Therefore, examining rsid rs80043958 at chr5:88,179,576 with an IGAP GWAS pvalue of
0.002, we see the ADNI heterozygosity is 0.269 - this and other information about that variant is
displayed in a popup by clicking on the pinkish "anchor block" - keep in mind that that number would
be more informative if broken out by phenotype.

```{r adniVCF,  results='hide'}
vcfFilename <- system.file(package="igvR", "extdata", "mef2c-4kb.vcf")
vcf <- readVcf(vcfFilename, "hg19")
track.vcf <- VariantTrack("AMPAD VCF", vcf, trackHeight=1000)
Sys.sleep(5)
displayTrack(igv, track.vcf)
Sys.sleep(10)  # allow plenty of time for this to complete, and the track to appear, before ending this chunk
```
# References

1. Lambert, Jean-Charles, et al. "Meta-analysis of 74,046 individuals identifies 11 new susceptibility loci for Alzheimer's disease." [Nature genetics 45.12 (2013): 1452](https://www.ncbi.nlm.nih.gov/pubmed/24162737).

```{r sessionInfo}
sessionInfo()
````

